### ISSUE TITLE
Memory grows and never released

### ISSUE THREAD

**Issue by abdessamad-zevenet** [openssl/openssl#8993]
I'm developing a simple multi threaded HTTP/s server using non-blocking socket to handle HTTP request. With every new connection  the memory keep growin  and I can't figure out the reason. It seems that it related to SSL code because it doesn't happen when using simple HTTP request.
Bellow is a simplified version of the most important part of code related to SSL operation (may contain errors and no error checking).
 Using valgrind, heaptrack and leak sanitizers doesn't help since it seems that the memory is still reachable and no leak was detected.
Any help on what I'm missing ?
How would be the proper way and order to clean Connection struct SSL and BIO data ? (freeSsl() ).

NOTES:
- Tried Openssl 1.1 and 1.1.1 from Debian and self build, they behave the same.
- Memory growth is more noticeable when clients use TLS 1.2
- Tried disabling SESSION caching and setting SSL_MODE_RELEASE_BUFFERS

```
struct SSLContext{
	SSL_CTX *ssl_ctx{nullptr};
	SSLContext(){
		ERR_load_crypto_strings();
		ERR_load_SSL_strings();
		SSL_load_error_strings();
		OpenSSL_add_all_algorithms();
		int r = SSL_library_init();
		ssl_ctx = SSL_CTX_new(SSLv23_method());
		int r = SSL_CTX_use_certificate_file(ssl_ctx, global::cert_file.c_str(),
			SSL_FILETYPE_PEM); 
		r = SSL_CTX_use_PrivateKey_file(ssl_ctx, global::key_file.c_str(), SSL_FILETYPE_PEM); 
		r = SSL_CTX_check_private_key(ssl_ctx); 
	}

	virtual ~SSLContext(){
		SSL_CTX_free(ssl_ctx);
		ERR_free_strings();
	}
};
```

```
struct Connection
{
	int sock_fd;
	SSL *ssl{nullptr}; 
  	BIO *sbio{nullptr};// socket bio  
  	BIO *io{nullptr};// buffer bio  
  	BIO *ssl_bio{nullptr};// ssl bio
	~Connection(){ //free ssl objects
		if (ssl != nullptr) {
			SSL_set_shutdown(ssl, SSL_RECEIVED_SHUTDOWN);
			freeSsl();
			if(sock_fd > 0) ::close(sock_fd);
		}
	}
	void freeSsl(){
		SSL_shutdown(ssl);
		SSL_free(ssl);
		if(io != nullptr) {
			BIO_flush(io);
			BIO_free(io);
			io = nullptr;
		}
		if(ssl_bio != nullptr) {
			BIO_flush(ssl_bio);
			BIO_free(ssl_bio);     
			ssl_bio = nullptr; 
		}
	}
	bool enableReadEvent(); //enable socket read events, set EPOLLIN | EPOLLET
	bool enableWriteEvent();//enable socket write event, set EPOLLOUT | EPOLLET	
};

```

```
struct SSLConnectionManager
{
	static SSLContext ssl_context; //shared between all SSL workers

	bool handleHandshake(Connection &ssl_connection)
	{
		if(ssl_connection.ssl != nullptr) ssl_connection.freeSsl();		
		ssl_connection.ssl = SSL_new(ssl_context->ssl_ctx);
		ssl_connection.sbio = BIO_new_socket(ssl_connection.sock_fd, BIO_CLOSE);
		SSL_set_bio(ssl_connection.ssl, ssl_connection.sbio, ssl_connection.sbio);
		ssl_connection.io = BIO_new(BIO_f_buffer());
		ssl_connection.ssl_bio = BIO_new(BIO_f_ssl());
		BIO_set_ssl(ssl_connection.ssl_bio, ssl_connection.ssl, BIO_CLOSE);
		BIO_push(ssl_connection.io, ssl_connection.ssl_bio);
		SSL_set_accept_state(ssl_connection.ssl);

		int r = SSL_do_handshake(ssl_connection.ssl); 
		if (r == 1) {
			ssl_connection.ssl_connected = true;   
			ssl_connection.enableReadEvent();
			return true;
		}

		int err = SSL_get_error(ssl_connection.ssl, r);
		if (err == SSL_ERROR_WANT_WRITE || err == SSL_ERROR_WANT_READ) {  
			ssl_connection.enableReadEvent();
			return true;
		} 
	    //"SSL_do_handshake error, abort connection
		return false;    

	}

	IO::IO_RESULT SSLConnectionManager::handleRead(Connection &ssl_connection) {
		if (!ssl_connection.ssl_connected) {
			return IO::IO_RESULT::SSL_NEED_HANDSHAKE;
		}
		int rc = -1;
		int bytes_read = 0;
		for (;;) {
			rc = BIO_read(ssl_connection.io,
				ssl_connection.buffer + ssl_connection.buffer_size,
				static_cast<int>(MAX_DATA_SIZE - ssl_connection.buffer_size));        
			if (rc == 0) {
				return bytes_read > 0 ? IO::IO_RESULT::SUCCESS : IO::IO_RESULT::ZERO_DATA_READ;            
			}else if (rc < 0) {
				if (BIO_should_retry(ssl_connection.io)) {
					return IO::IO_RESULT::DONE_TRY_AGAIN;
				}
				return IO::IO_RESULT::ERROR;
			}
			bytes_read += rc;
			ssl_connection.buffer_size += static_cast<size_t>(rc);
		}
		return IO::IO_RESULT::SUCCESS;
	}

	IO::IO_RESULT SSLConnectionManager::handleWrite(Connection &ssl_connection,
		const char *data, size_t data_size, size_t &written) {
		if (!ssl_connection.ssl_connected) {
        	return IO::IO_RESULT::SSL_NEED_HANDSHAKE; // after we call handleHanshake
        }

        IO::IO_RESULT result;
        int rc = -1;
        written = 0;
        for (;;) {
        	rc = BIO_write(ssl_connection.io, data + written, static_cast<int>(data_size - written));        
        	if (rc == 0) {
        		result = IO::IO_RESULT::DONE_TRY_AGAIN;
        		break;
        	} else if (rc < 0) {
        		if (BIO_should_retry(ssl_connection.io)) {
        			result = IO::IO_RESULT::DONE_TRY_AGAIN;    			
        			break;
        		} else {
        			return IO::IO_RESULT::ERROR;    				
        		}
        	} else {
        		written += rc;
        		if ((data_size - written) == 0) {
        			result = IO::IO_RESULT::SUCCESS;
        			break;
        		};
        	}
        }
        BIO_flush(ssl_connection.io);  
        return result;
    }	
};
```

```
struct SSLWorker : EpollManager{	

	SSLConnectionManager ssl_connection_manager;

	bool onConnectEvent(Connection &ssl_connection){
		auto sock_fd  = accept(...);	
		Connection * new_connection = new  Connection(sock_fd);
		addToEventManager(*new_connection, EV_READ);
	} 

	void doWork() {
		is_running = true;
		int res = 0;
		epoll_event events[1024];
		while (is_running) {
			res = ::epoll_wait(epoll_fd, events, 1024, -1);
			if (res < 0)
				return;
			for (int i = 0; i < res; i++) {
				auto conn = static_cast<Connection *>(events[i].data.ptr);
				if (events[i].events & (EPOLLHUP | EPOLLERR | EPOLLRDHUP)) {
					delete conn; //remote closed connection, free all
				} else {
					if (events[i].events & EPOLLIN) {
						if (conn->fd == global::listen_fd) {            
							onConnectEvent();
							continue;
						} else {            
							ssl_connection_manager.handleRead(*conn);
							processRequest();
							conn.enableWriteEvent();
						}
					}
					if (events[i].events & EPOLLOUT) {   
						ssl_connection_manager.handleWrite(*conn);
					}
				}
			}
		}
	}


};
```

**Comment by abdessamad-zevenet** [comment]
In brief execution, the only leaking stack that is detected by heaptrack.
![Screenshot from 2018-11-29 17-44-19](https://user-images.githubusercontent.com/33321464/58489655-fe555600-816b-11e9-85fe-7517ce655132.png)

**Comment by Geal** [comment]
I think I see a similar issue, used with the rust-openssl library. I don't know if I'll be able to write a reduced test case that reproduces it, but I can try.
Some information that might be useful:
- memory leak detection in valgrind or leak sanitizer have not found anything yet
- memory usage can be very stable for hours, but once the leak starts, it leaks very regularly. cf the picture, each line is a different process
<img width="796" alt="Screenshot 2019-09-20 at 15 40 37" src="https://user-images.githubusercontent.com/119296/65712835-31ffa880-e098-11e9-9b6d-5065f71bb5c0.png">
- still leaking with 1.1.1d
Are there things I should look for, or tests I should run? I can recompile an openssl version if needed

**Comment by abdessamad-zevenet** [comment]
If no memory leak is detected, first I would try to release free heap blocks back to the system using malloc_trim() libc function in case memory is being pooled (most likely) to how much exactly is being used.
Also, make sure to free SSL resources correctly, in my case memory usage improved a lot after setting BIO's _NOCLOSE and freeing everything manually.

**Comment by nhorman** [comment]
Given that (a) the code above is not in a state where it can be made buildable and (b) our asan test runs in CI show not leaks, I'm going to assume this issue has been fixed sometime in the last 5 years.

Marking as inactive, to be closed at the end of 3.4 dev, barring further input
